internal/launcher/launcher.go

```diff
@@ -0,0 +1,158 @@
+// Package launcher is the entry point for the k6 command.
+package launcher
+
+import (
+	"context"
+	"errors"
+	"os/exec"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/grafana/k6deps"
+	"go.k6.io/k6/cmd/state"
+	"go.k6.io/k6/internal/build"
+	k6Cmd "go.k6.io/k6/internal/cmd"
+)
+
+// Execute runs the k6 command.
+func Execute() {
+	gs := state.NewGlobalState(context.Background())
+
+	newLauncher(gs).launch()
+}
+
+// launcher is a k6 launcher
+type launcher struct {
+	gs *state.GlobalState
+	// function to fall back if binary provisioning is not required
+	fallback func(gs *state.GlobalState)
+	// function to provision a k6 binary that satisfies the dependencies
+	provision func(*state.GlobalState, k6deps.Dependencies) (string, string, error)
+	// function to execute k6 binary
+	run func(*state.GlobalState, string) (int, error)
+}
+
+func newLauncher(gs *state.GlobalState) *launcher {
+	return &launcher{
+		gs:        gs,
+		fallback:  k6Cmd.ExecuteWithGlobalState,
+		provision: k6buildProvision,
+		run:       runK6Cmd,
+	}
+}
+
+// launch executes k6 either by launching a provisioned binary or defaulting to the
+// current binary it this is not necessary.
+// If the fhe fallback is called, it can exit the process so don't assume it will return
+func (l *launcher) launch() {
+	// if binary provisioning not enabled, continue with regular k6 execution path
+	if !l.gs.Flags.BinaryProvisioning {
+		l.gs.Logger.Debug("binary provisioning disabled")
+		l.fallback(l.gs)
+		return
+	}
+
+	l.gs.Logger.Info("trying to provision binary")
+
+	deps, err := k6deps.Analyze(newDepsOptions(l.gs, l.gs.CmdArgs[1:]))
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("failed to analyze dependencies, can't try binary provisioning, please report this issue")
+		l.gs.OSExit(1)
+	}
+
+	// binary provisioning enabled but not required by this command
+	// continue with regular k6 execution path
+	if !isCustomBuildRequired(build.Version, deps) {
+		l.gs.Logger.
+			Debug("binary provisioning not required")
+		l.fallback(l.gs)
+		return
+	}
+
+	l.gs.Logger.
+		WithField("deps", deps).
```

Comment:
Are we sure of the print here? Is it safe? Is it a nice and understandable print?